text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,401 @@
+
+    #[wasm_bindgen(method, structural, js_name = removeEventListener)]
+    fn remove_event_listener(this: &JsValue, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool);
+
+    #[wasm_bindgen(method, structural, js_name = dispatchEvent)]
+    fn dispatch_event(this: &JsValue, event: Event);
+}
+
+pub trait IEventTarget: AsRef<JsValue> {
+    #[inline]
+    fn add_event_listener(&self, name: &str, listener, &Closure<FnMut(Event)>, use_capture: bool) {
+        add_event_listener(self.as_ref(), name, listener, use_capture)
+    }
+
+    #[inline]
+    fn remove_event_listener(&self, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool) {
+        remove_event_listener(self.as_ref(), name, listener, use_capture)
+    }
+
+    #[inline]
+    fn dispatch_event(&self, event: Event) {
+        dispatch_event(self.as_ref(), event)
+    }
+}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(method, structural, getter = nodeName)]
+    fn node_name(this: &JsValue) -> JsString;
+
+    #[wasm_bindgen(method, structural, getter = textContent)]
+    fn text_content(this: &JsValue) -> JsString;
+
+    #[wasm_bindgen(method, structural, js_name = appendChild)]
+    fn append_child<A: INode>(this: &JsValue, child: A) -> A;
+
+    #[wasm_bindgen(method, structural, js_name = removeChild)]
+    fn remove_child<A: INode>(this: &JsValue, child: A) -> A;
+}
+
+pub trait INode: IEventTarget {
+    #[inline]
+    fn node_name(&self) -> JsString {
+        node_name(self.as_ref())
+    }
+
+    #[inline]
+    fn text_content(&self) -> JsString {
+        text_content(self.as_ref())
+    }
+
+    #[inline]
+    fn append_child<A: INode>(&self, child: A) -> A {
+        append_child(self.as_ref(), child)
+    }
+
+    #[inline]
+    fn remove_child<A: INode>(&self, child: A) -> A {
+        remove_child(self.as_ref(), child)
+    }
+
+    // And a bunch more methods...
+}
+```
+
+As you can see, it moves the trait methods into an `extern` (which has a `wasm-bindgen` attribute).
+
+Each method that takes `&self` or `&mut self` gets translated into a `method, structural` extern.
+
+Everything else is moved as-is from the trait to the extern.
+
+Then the trait methods are rewritten so that they call the extern functions (using `as_ref()` to convert `&self` to a `&JsValue`),
+and they are marked as `#[inline]`.
+
+Now that we have defined the desired traits and methods, we can add the methods to a type like this:
+
+```rust
+#[wasm_bindgen]
+extern {
+    type EventTarget;
+}
+
+impl IEventTarget for EventTarget {}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    type Node;
+}
+
+impl IEventTarget for Node {}
+impl INode for Node {}
+```
+
+Because the traits contain default implementations, this is all that is needed to make it work.
+
+## Mixins
+
+The above technique works great for class-based inheritance, but it can also be used to support mixins.
```

Comment:
The WebIDL spec explicitly says that mixins are not meant to be exposed in language bindings, but are merely an editorial tool. In particular, for the same reason as just stuffing everything onto JsValues, this could be incompatible with the host-bindings proposal.

Reply:
Since the methods are a part of a mixin and not a class, how will that work with the host-bindings proposal?

Reply:
Mixins are a shorthand way of including methods and/or attributes in multiple interfaces. 

Consider the webgl vs webgl2 contexts. Many of the methods are the same, but there is no real relation between them from a class hierarchy sense. To simplify spec specification, you can use place all the redundant declarations in a mixin and then include it for each of the contexts. This is the same as just copying and pasting all of the redundant declarations into each of their interface declarations.

We already support mixins (and partial interfaces, which are similar in concept) in our webidl crate, so I would really just remove the section on mixins from the RFC.